numba_dppy/utils/array_utils.py

```diff
@@ -0,0 +1,205 @@
+    try:
+        usm_mem = dpctl_mem.as_usm_memory(obj)
+    except:
+        if hasattr(obj, "base"):
+            try:
+                usm_mem = dpctl_mem.as_usm_memory(obj.base)
+            except Exception:
+                pass
+
+    return usm_mem
+
+
+def copy_to_usm_backed(usm_backed, obj):
+    """
+    Copy from supported objects to USM backed data.
+
+    Args:
+        usm_backed: Object conformant to __sycl_usm_array_interface__.
+        obj (numpy.ndarray): Numpy ndarray, the data will be copied into.
+
+    Raises:
+        TypeError: If any argument is not of permitted type.
+        ValueError: If size of data does not match.
+    """
+    usm_mem = is_usm_backed(usm_backed)
+    if usm_mem is None:
+        raise TypeError("Source is not USM backed.")
+
+    if not isinstance(obj, np.ndarray):
+        raise TypeError(
+            "Obj is not USM backed and is not of type "
+            "numpy.ndarray. Obj type: %s" % (type(obj))
+        )
+
+    if obj.dtype not in [np.dtype(typ) for typ in supported_numpy_dtype]:
+        raise ValueError(
+            "dtype is not supprted. Supported dtypes "
+            "are: %s" % (supported_numpy_dtype)
+        )
+
+    size = np.prod(obj.shape)
+    if usm_mem.size != (obj.dtype.itemsize * size):
+        raise ValueError(
+            "Size (Bytes) of data does not match. USM backed "
+            "memory size %d, supported object size: %d"
+            % (usm_mem.size, (obj.dtype.itemsize * size))
+        )
+
+    obj_memview = memoryview(obj)
+    obj_memview = obj_memview.cast("B")
+    usm_mem.copy_from_host(obj_memview)
+
+
+def copy_from_usm_backed(usm_backed, obj):
+    """
+    Copy from USM backed data to supported objects.
+
+    Args:
+        usm_backed: Object conformant to __sycl_usm_array_interface__.
+        obj (numpy.ndarray): Numpy ndarray, the data will be copied into.
+
+
+    Raises:
+        TypeError: If any argument is not of permitted type.
+        ValueError: If size of data does not match.
+    """
+    usm_mem = is_usm_backed(usm_backed)
+    if usm_mem is None:
+        raise TypeError("Source is not USM backed.")
+
+    if not isinstance(obj, np.ndarray):
+        raise TypeError(
+            "Obj is not USM backed and is not of type "
+            "numpy.ndarray. Obj type: %s" % (type(obj))
+        )
+
+    if obj.dtype not in [np.dtype(typ) for typ in supported_numpy_dtype]:
+        raise ValueError(
+            "dtype is not supprted. Supported dtypes "
+            "are: %s" % (supported_numpy_dtype)
+        )
+
+    size = np.prod(obj.shape)
+    if usm_mem.size != (obj.dtype.itemsize * size):
+        raise ValueError(
+            "Size (Bytes) of data does not match. USM backed "
+            "memory size %d, supported object size: %d"
+            % (usm_mem.size, (obj.dtype.itemsize * size))
+        )
+
+    obj_memview = memoryview(obj)
+    obj_memview = obj_memview.cast("B")
+    usm_mem.copy_to_host(obj_memview)
```

Comment:
`copy_to_host` can throw a `ValueError` that should handled.

Reply:
Can you please elaborate on how the handling should be done? My understanding is `copy_to_host()` will raise whatever Error it finds and we do not need to reraise it here. Please correct me if there is a misunderstanding on my part.